packages/suite/src/actions/suite/metadataLabelingActions.ts

```diff
@@ -529,6 +568,297 @@ export const addMetadata =
+        }
+
+        dispatch(syncMetadataKeys(device, prevEncryptionVersion));
+
+        if (!device.state) {
+            // this should never happen
+            return { success: false, error: 'metadata migration: device not authorized' };
+        }
+
+        // 8. get labelable entitites again (there was async operation in between)
+        const allEntities = dispatch(getLabelableEntities(device.state));
+
+        // 9. split labelable entities into 2 groups:
+        //    - entitiesToMigrate: don't have new file && have old file
+        //    - entititiesToCreateDummies: don't have new file && don't have old file
+
+        const entitiesToMigrate: LabelableEntity[] = [];
+        const entititiesToCreateDummies: LabelableEntity[] = [];
+        allEntities.forEach(entity => {
+            const prevKeys = entity[prevEncryptionVersion];
+
+            if (!prevKeys) {
+                console.error('metadata migration: prev keys are missing');
+
+                return; // should never happen
+            }
+            const nextKeys = entity[2];
+
+            if (!nextKeys) {
+                console.error('metadata migration: next keys are missing');
+
+                return; // should never happen
+            }
+
+            const oldFileExists = oldEncryptionFiles.find(file => file === prevKeys.fileName);
+            const newFileExists = currentEncryptionFiles.find(file => file === nextKeys.fileName);
+
+            if (newFileExists) {
+                return; // already migrated
+            }
+
+            if (!oldFileExists) {
+                entititiesToCreateDummies.push(entity);
+
+                return; // there is nothing to migrate,
+            }
+
+            entitiesToMigrate.push(entity);
+        });
+
+        // 10. now all data is ready. we know what operations should be carried out. dummy files will be created, old files will be migrated and their content will be filled into local state
+
+        // NOTE: I understand that this is not the right layer to rate limit access to provider API. It should be handled in provider service itself but
+        // I don't have free hands to do it now. So I am running all requests in series as a workaround now. Correct solution would be
+        // implementing provider.batchWrite and do batching if possible and if not, use single requests with some rate limiting
+        const promises = [
+            ...entitiesToMigrate.map(
+                entity => () =>
+                    dispatch(createMigrationPromise(entity, prevEncryptionVersion, true, device!)),
+            ),
+
+            // todo: maybe not needed to create dummies in case we manage to migrate all old files.
+            ...entititiesToCreateDummies.map(
+                entity => () =>
+                    dispatch(createMigrationPromise(entity, prevEncryptionVersion, false, device!)),
+            ),
+        ];
+
+        for (let i = 0; i < promises.length; ++i) {
```

Comment:
nit/optional: we may introduce some helper `await promiseResolveSequentially(promises)`